modules/cuml/src/umap.ts

```diff
@@ -0,0 +1,111 @@
+import {
+  DataFrame,
+  Float32,
+  Numeric,
+  Series,
+  TypeMap,
+} from '@rapidsai/cudf';
+import {DeviceBuffer} from '@rapidsai/rmm/';
+import {CUMLLogLevels, MetricType} from './mappings';
+
+import {UMAPBase, UMAPInterface, UMAPParams} from './umap_base';
+import {series_to_dataframe, transform_input_to_device_buffer} from './utilities/array_utils';
+
+export class UMAP {
+  public _umap: UMAPInterface;
+  public _embeddings: DeviceBuffer;
+
+  protected constructor(input: UMAPParams) {
+    this._umap       = new UMAPBase(input);
+    this._embeddings = new DeviceBuffer();
+  }
+
+  private _generate_embeddings(n_samples: number): Series<Float32> {
+    return Series.sequence(
+      {type: new Float32, size: n_samples * this._umap.nComponents, init: 0, step: 0});
+  }
+
+  fit(X: Series<Numeric>|DataFrame<TypeMap>,
+      y: null|Series<Numeric>|DataFrame<TypeMap>,
+      convertDType: boolean): void {
+    const n_samples  = (X instanceof Series) ? X.length : X.numRows;
+    const n_features = (X instanceof Series) ? 1 : X.numColumns;
+    this._embeddings = transform_input_to_device_buffer(this._generate_embeddings(n_samples));
+
+    this._embeddings = this._umap.fit(transform_input_to_device_buffer(X),
+                                      n_samples,
+                                      n_features,
+                                      (y == null) ? null : transform_input_to_device_buffer(y),
+                                      null,
+                                      null,
+                                      convertDType,
+                                      this._embeddings);
+    return undefined;
+  }
```

Comment:
This method looks difficult to strongly type based on its parameters. Can we split this into separate methods based on input types? We can still have a private implementation once the inputs have been converted into a DeviceBuffer matrix.